design/artifacts-in-pipelines.md

```diff
@@ -23,103 +19,414 @@ Pipelines should support customers invoking a supported (see later section) tool
+
+#### Edge cases (for documentation purposes)
+
+The only way to change the location of `settings.xml` is via the [`mvn -s` parameter](https://stackoverflow.com/questions/25277866/maven-command-line-how-to-point-to-a-specific-settings-xml-for-a-single-command). Thus, Maven users running multiple agents on the same user profile will need to be aware that builds could clash. These users should use the documentation (see last section) that covers how to manually construct a `settings.xml` using `SYSTEM_ACCESSTOKEN` and pass it using `mvn -s`.
+
+#### Example
+
+```yaml
+- task: MavenAuthenticate@0
+  inputs:
+    artifactsFeeds: codesharing-demo
+    mavenServiceConnections: myMavenFeed
+```
+
+#### Task reference
+
+```yaml
+# Maven Authenticate
+- task: MavenAuthenticate@0
+  inputs:
+    artifactsFeeds: # Optional, one or more Azure Artifacts feed names/GUIDs, comma-separated
+    mavenServiceConnections: # Optional, one or more Maven-typed service connection names/GUIDs, comma-separated
+```
+
+#### Open questions
+
+- Should we put envvars in the settings.xml, [like so](https://stackoverflow.com/questions/31251259/how-to-pass-maven-settings-via-environmental-vars)? Seems unnecessary since the settings.xml is already outside the workspace.
+- Is "comma-separated" the norm for multiple inputs in YAML?
 
 ### npm
 
+consider: ever generate npmrc?
+
 The `npm` utility (which comes with Node.js) is used to install packages and tools required for your Node.js development. `npm` supports Azure Artifacts feeds and external feeds using basic or token authentication.
 
-For npm, the `use` syntax will create or update an `.npmrc` file at the root of the repo with credentials from the provided feed/service connection.
+#### Inputs
+
+- 0-1 Azure Artifacts feed name/GUID
+- 0+ npm-typed service connection names/GUIDs
+- Agent's proxy information, if available
+
+#### Actions
+
+##### Authentication
+
+The task looks for an `.npmrc` either in the root of the repo or at the user's provided location and sets appropriate [environment variables](https://docs.npmjs.com/misc/config#environment-variables) for each item provided. For example, if a user provids a service connection for npmjs with an authtoken, the task will set an environment variable:
+
+`npm_config_//registry.npmjs.org/:_authToken=token`
+
+If the task discovers or if a user provides an Azure Artifacts feed name, the task will set 3 environment variables:
+
+`npm_config_//pkgs.dev.azure.com/codesharing-demo/_packaging/codesharing-demo/npm/registry/:username=codesharing-demo` 
+
+`npm_config_//pkgs.dev.azure.com/codesharing-demo/_packaging/codesharing-demo/npm/registry/:_password=token-here`
+
+`npm_config_//pkgs.dev.azure.com/codesharing-demo/_packaging/codesharing-demo/npm/registry/:email=npm requires email to be set but doesn't use the value`
+
+##### Proxy
+
+If the agent's proxy information is available, the task will set the `HTTPS_PROXY` and/or `HTTP_PROXY` variables as appropriate.
```

Comment:
How about NO_PROXY? This one is harder, because it's not just a value we can plumb through, so it would be good to document what it will get set to. We have a list of regexes of the hosts that should be bypassed, so at the moment in the tasks the registries found in the npmrc file are run against the regexes, and if there are any matches, those registries are added to NO_PROXY.